### PR TITLE
Added poweroff_guest and reboot_guest

### PR DESCRIPTION
**Pull request by Heusini**
##### SUMMARY
Added the option to poweroff_guest and reboot_guest so you can make gracefull VM shutdowns or reboots 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
modules/cloud/vmware/vsphere_guest.py

##### ANSIBLE VERSION
```
devel
```

### CODE REVIEW

**Comment by Akasurde** [comment]
@Heusini Thanks for the contribution. We planning to deprecate `vsphere_guest` in favor of `vmware_guest`. I would recommend to use `vmware_guest` module which already provides this functionality.

Please see documentation about `vsphere_guest` deprecation [here](https://github.com/ansible/community/wiki/VMware:-vsphere_guest_deprecation)

Let me know if you need any help from my side.

**Comment by Akasurde** [comment]
@Heusini Are you still working on this ? also, 
Could you please add `-vvvv` output here so that we can take it forward?

In one of the IRC meeting of VMware SIG, we decided to merge all vsphere_guest PRs based on `-vvv` output. This decision was taken as vsphere_guest is deprecated in version 2.5. You can find the details [here](https://github.com/ansible/community/wiki/VMware%3A-vsphere_guest_deprecation)

Let me know if you need any other information.

**Comment by Heusini** [comment]
Hi no i dont work on this anymore we moved to vmware_guest. But here is the output; hope i did it right.

**Playbook:**
```yaml
---
tasks:
      - name: vmware_guest_restart
        vsphere_guest:
          vcenter_hostname: "{{ vcenter }}"
          username: "{{ vcenter_user }}"
          password: "{{ vcenter_pass }}"
          guest: "{{ config.name }}"
          validate_certs: no
          state: reboot_guest
        delegate_to: localhost
```

**Output: -vvvv**
```
---
[user@hostname ansible]# ansible-playbook pb_restart_gracefull.yml --limit xxxx -vvvv
ansible-playbook 2.4.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
Using /etc/ansible/ansible.cfg as config file
setting up inventory plugins
Parsed /etc/ansible/external-inventory/vmware_inv.py inventory source with script plugin
Parsed /etc/ansible/external-inventory/hosts inventory source with ini plugin
Parsed /etc/ansible/external-inventory/vmware_inv_2.py inventory source with script plugin

Loading callback plugin default of type stdout, v2.0 from /usr/lib/python2.7/site-packages/ansible/plugins/callback/__init__.pyc

PLAYBOOK: pb_restart_gracefull.yml *************************************************************************************************************************************************************************
1 plays in pb_restart_gracefull.yml

PLAY [reboot hosts] ****************************************************************************************************************************************************************************************
META: ran handlers

TASK [vmware_guest_restart] *********************************************************************************************************************************************************************************
task path: /etc/ansible/pb_restart_gracefull.yml:11
Using module file /etc/ansible/library/vsphere_guest.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1519121509.98-190501915797727 `" && echo ansible-tmp-1519121509.98-190501915797727="` echo /root/.ansible/tmp/ansible-tmp-1519121509.98-190501915797727 `" ) && sleep 0'
<localhost> PUT /tmp/tmpEctoBV TO /root/.ansible/tmp/ansible-tmp-1519121509.98-190501915797727/vsphere_guest.py
<localhost> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1519121509.98-190501915797727/ /root/.ansible/tmp/ansible-tmp-1519121509.98-190501915797727/vsphere_guest.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python2 /root/.ansible/tmp/ansible-tmp-1519121509.98-190501915797727/vsphere_guest.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1519121509.98-190501915797727/" > /dev/null 2>&1 && sleep 0'
changed: [xxxx -> localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "cluster": null,
            "esxi": {},
            "force": false,
            "from_template": null,
            "guest": "xxxx",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "power_on_after_clone": true,
            "resource_pool": null,
            "snapshot_to_clone": null,
            "state": "reboot_guest",
            "template_src": null,
            "username": "usr",
            "validate_certs": false,
            "vcenter_hostname": "vchost",
            "vm_disk": {},
            "vm_extra_config": {},
            "vm_hardware": {},
            "vm_hw_version": null,
            "vm_nic": {},
            "vmware_guest_facts": null
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************************************************************************************************************************
xxxx : ok=1    changed=1    unreachable=0    failed=0
```

**Comment by Akasurde** [comment]
@Heusini Can you please rebase the branch so that I can merge this ?

File: lib/ansible/modules/cloud/vmware/vsphere_guest.py [hunk 1]
```diff
@@ -1826,7 +1845,7 @@ def main():
                 module.fail_json(msg="Fact gather failed with exception %s"
                                  % to_native(e), exception=traceback.format_exc())
         # Power Changes
-        elif state in ['powered_on', 'powered_off', 'restarted']:
+        elif state in ['powered_on', 'powered_off', 'restarted','reboot_guest', 'poweroff_guest']:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vsphere_guest.py]: space after ,

File: lib/ansible/modules/cloud/vmware/vsphere_guest.py [hunk 2]
```diff
@@ -1553,10 +1553,27 @@ def power_state(vm, state, force):
...
+                else:
+                    return "Cannot poweroff_guest VM in the current state %s" \
+                        % power_status
 
             elif state == 'powered_on':
                 vm.power_on(sync_run=True)
 
+            elif state == 'reboot_guest':
+                if power_status in ('POWERED ON'):
+                      vm.reboot_guest()
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vsphere_guest.py]: over indentation

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by willthames** [comment]
rebased this as it made it look as there are is a single PR open on a lot of files.

**Review by Akasurde — APPROVED** [review]

**Pull request merged**